lib/src/repo.rs

```diff
@@ -243,30 +245,39 @@ impl ReadonlyRepo {
         })
     }
 
-    fn commit_change_id_index(&self) -> &IdIndex {
-        self.commit_change_id_index.get_or_init(|| {
+    fn commit_id_index(&self) -> &IdIndex {
+        self.commit_id_index.get_or_init(|| {
+            // We need to account for rewritten commits as well
+            let mut id_index = IdIndex::new();
+            for entry in self.as_repo_ref().index().iter() {
+                id_index.insert(entry.commit_id().hex().as_bytes(), ());
+            }
+            id_index
+        })
+    }
+
+    fn change_id_index(&self) -> &IdIndex {
+        self.change_id_index.get_or_init(|| {
             let all_visible_revisions = crate::revset::RevsetExpression::all()
                 .evaluate(self.as_repo_ref(), None)
                 .unwrap();
-            let change_hex_iter = all_visible_revisions
-                .iter()
-                .map(|index_entry| index_entry.change_id().hex());
-            // We need to account for rewritten commits as well
-            let index = self.as_repo_ref().index();
-            let commit_hex_iter = index
-                .iter()
-                .map(|index_entry| index_entry.commit_id().hex());
             let mut id_index = IdIndex::new();
-            for id_hex in itertools::chain(change_hex_iter, commit_hex_iter) {
-                id_index.insert(id_hex.as_bytes(), ());
+            for entry in all_visible_revisions.iter() {
+                id_index.insert(entry.change_id().hex().as_bytes(), ());
             }
             id_index
         })
     }
 
     pub fn shortest_unique_prefix_length(&self, target_id_hex: &str) -> usize {
-        self.commit_change_id_index()
-            .shortest_unique_prefix_len(target_id_hex.as_bytes())
+        // For `len = index.shortest(unknown_id)`, `unknown_id[..len] not in index` is
+        // guaranteed. So taking `max()` should just work.
+        cmp::max(
+            self.commit_id_index()
+                .shortest_unique_prefix_len(target_id_hex.as_bytes()),
+            self.change_id_index()
+                .shortest_unique_prefix_len(target_id_hex.as_bytes()),
+        )
```

Comment:
It's documented as follows (or I tried to):
```
For `len = index.shortest(unknown_id)`, `unknown_id[..len] not in index` is
guaranteed. So taking `max()` should just work.
```

`shortest_unique_prefix_len()` behaves as if the given id exists in the index. If it doesn't (actually) exist in the index, the prefix is guaranteed to not match any ids in that space. Hence `max()` works.

Reply:
I think this is correct as is. The comment could say

    If `len = index.shortest_unique_prefix_len(id)`, then a prefix of length `len` will disambiguate it from
    all other ids in the index. This will be just as true for max(len, anything_else). So, a max of two such
    lengths disambiguates in both indexes.

Reply:
Nice! I've updated the inline comment, thanks. I use abstract `index.shortest(id)` form because I'll switch the commit_id index to slightly different API.

Reply:
I'll mark this as resolved per @ilyagr's comment. Thanks again for the suggestion.